shared_stack.bash

```diff
@@ -0,0 +1,214 @@
+#   - remove the $ROOT/tag subdirectory
+#
+# A -n option may be used to do a "dry run" that only prints out the commands
+# that would be executed.
+#
+# This code is intended to work properly if invoked multiple times in parallel,
+# e.g. if it takes longer than the interval between cron jobs, although there
+# is no resulting speedup.  Hence PID-labeled temporary directories and atomic
+# moves are used frequently.
+
+###############
+# CONFIGURATION
+###############
+
+# Software root directory
+ROOT=/scratch/rubinsw
+# State directory
+STATE=~/shared-stack
+# Top-level product to install
+PRODUCT=lsst_distrib
+# Number of daily releases to keep
+# shellcheck disable=SC2034
+KEEP_d=24
+# Number of weekly releases to keep
+# shellcheck disable=SC2034
+KEEP_w=26
+
+#######
+# Setup
+#######
+
+while getopts n opt; do
+  case "$opt" in
+    n) dryrun="echo"
+       ;;
+    *) echo "Usage: $0 [-n]"
+       exit 1 ;;
+  esac
+done
+[ $OPTIND -le $# ] && echo "Usage: $0 [-n]" && exit 1
+# Ensure that there's no contamination from a pre-existing environment
+unset LSST_CONDA_ENV_NAME PYTHONPATH
+
+##############
+# Installation
+##############
+
+set -e
+mkdir -p "$STATE"
+cd "$STATE"
+# Label temporaries with our unique PID
+tmp=tmp.$$
+mkdir -p "$tmp" "$ROOT/tag"
+[ -e tags.deleted ] || touch tags.deleted
+umask 022
+
+# Get list of tags
+$dryrun curl -L https://ls.st/lsstinstall -o $tmp/lsstinstall
+$dryrun mv $tmp/lsstinstall lsstinstall
+curl -L https://eups.lsst.codes/stack/src/tags | while read -r line; do
+  [[ "$line" =~ \>([dvw][0-9_]+(rc[0-9]+)?)\.list\< ]] && echo "${BASH_REMATCH[1]}"
+done | sort > $tmp/tags
+
+for tag in $(comm -1 -3 <(sort tags.deleted) $tmp/tags); do
+  dir="$ROOT/tag/$tag"
+  [ -d "$dir" ] && continue
+  $dryrun mkdir -p "${dir}.$tmp" || continue
+  (
+     set -e
+     $dryrun cd "${dir}.$tmp"
+     $dryrun bash "$STATE/lsstinstall" -T "$tag" -p "$ROOT"/conda -S
+     set +x
+     $dryrun source loadLSST.sh
+     set -x
+     # Install the developer add-on packages if needed
+     # When rubin-env-developer becomes available, replace with this:
+     # conda list --json | grep rubin-env-developer > /dev/null \
+     #   || mamba install rubin-env-developer
+     if [ -n "$dryrun" ]; then
+       echo "conda list --json | grep mypy > /dev/null || mamba install ..."
+     else
+       # shellcheck disable=SC2046
+       conda list --json | grep mypy > /dev/null \
+         || mamba install -y -c conda-forge --no-update-deps $(cat "$STATE/developer.txt")
+     fi
+     $dryrun eups distrib install -t "$tag" "$PRODUCT"
+
+     # Rewrite load scripts and write useful files
+     $dryrun cp loadLSST.sh tmp.loadLSST.sh
+     if [ -n "$dryrun" ]; then
+       echo "echo . $ROOT/loadLSST.sh ... > loadLSST.sh"
+       echo "echo conda activate ... > switch.sh"
+       echo "echo ... > env_name"
+     else
+       echo "LSST_CONDA_ENV_NAME=$LSST_CONDA_ENV_NAME . \"$ROOT/loadLSST.sh\"" > loadLSST.sh
```

Comment:
This seems to always use the loadLSST.sh from the root (which is copied later) but what happens if the script changes to where this version could not work with a newer loadLSST.sh script

Reply:
The general idea is to keep loadLSST.sh as simple and trivial as possible and therefore always backward-compatible.  If we're doing a lot of stuff in loadLSST.sh, something is wrong.

Reply:
I agree that is ideal, but in a one off case where something needs to change? Do we just re-write affected environments, or perhaps not worry about them?

Reply:
We could always manually adjust a tag's loadLSST.sh if necessary, but with the stack physically embedded in the environment, it's hard for me to see what kind of changes would be possible.